### PR TITLE
Add OpenShift vendoring

### PR DESCRIPTION
**Pull request by kadel**
OpenShift patches some packages in its vendor directory. Those patched libs are not published elsewhere.
There is a proposal to fix it, but it is still discussed -  https://github.com/openshift/origin/pull/15797 

We have to use a modified version of those packages from github.com/openshift/orogin/vendor, otherwise, OpenShift won't build.

 `scripts/vendor-openshift.sh` - 
 This script adds OpenShift and all packages that are vendored by OpenShift to project vendor directory. It has to be run from root directory of the project (where vendor directory is).
You should run 'glide update -v' before running this script.

From now on, we should call `make vendor-update` instead of `glide update`

TODO:
- [x] Clean up commit messages and squash some commits
- [x] tests is complaining because currently, it is not possible to clean vendor with `glide-vc --only-code --no-tests --use-lock-file` (it won't build after that). It looks like we will have to stop using `glide-vc` :-(
- [x] add documentation to Development Guide

### CODE REVIEW

**Comment by kadel** [comment]
@cdrage how are we going to do this? Are you going to continue with DeploymentConfig support on this PR?

**Comment by cdrage** [comment]
@kadel Yes, I'll continue now on adding support for DeploymentConfig on PR: https://github.com/kedgeproject/kedge/pull/288

I'm assuming that by updating https://github.com/kedgeproject/kedge/pull/288 it will also update the commit in this PR? (if not, we could merge this and then continue in a different PR)

**Comment by kadel** [comment]
> I'm assuming that by updating #288 it will also update the commit in this PR? (if not, we could merge this and then continue in a different PR)

It won't :-(

But you should be able to modify my branch.

**Comment by cdrage** [comment]
@kadel Alright. I'll continue work on #288 until I have full support on it and then push it to this PR.

Thanks for all the vendoring fixes / work. So glad it's fixed now.

**Comment by containscafeine** [comment]
@kadel how about first `glide update`, then `glide-vc` and then copy OpenShift's vendor to our vendor.

**Comment by kadel** [comment]
@cdrage  now this PR is only about OpenShift vendoring

File: docs/development.md [hunk 1]
```diff
@@ -53,11 +53,10 @@ git push -f origin myfeature
...
 3. Check out the pull request process for more details
 
-## `glide`, `glide-vc` and dependency management
+## Dependency management
 
-Kedge uses `glide` to manage dependencies and `glide-vc` to clean vendor directory.
-They are not strictly required for building Kedge but they are required when managing dependencies under the `vendor/` directory.
-If you want to make changes to dependencies please make sure that `glide` and `glide-vc` are installed and are in your `$PATH`.
+Kedge uses `glide` to manage dependencies.
+If you want to make changes to dependencies please make sure that `glide`  are installed and are in your `$PATH`.
```
Comment (cdrage) [inline docs/development.md]: `is in your`

File: docs/development.md [hunk 2]
```diff
@@ -87,16 +84,14 @@ glide-vc -h
   Add new packages or subpackages to `glide.yaml` depending if you added whole
   new package as dependency or just new subpackage.
 
-2. Get new dependencies
-
-```bash
-glide update --strip-vendor
-```
+  Kedge vendors whole OpenShift and all its dependencies. (see comments in `glide.yaml` and `./scripts/vendor-openshift.yaml`)
```
Comment (cdrage) [inline docs/development.md]: remove `whole`

File: docs/development.md [hunk 3]
```diff
@@ -87,16 +84,14 @@ glide-vc -h
...
+  It is possible that dependence you want to add is already in vendor as OpenShift dependency. If that is true, please make sure
```
Comment (cdrage) [inline docs/development.md]: `the dependency`
`in OpenShift as a vendored dependency`

File: docs/development.md [hunk 4]
```diff
@@ -87,16 +84,14 @@ glide-vc -h
...
   new package as dependency or just new subpackage.
 
-2. Get new dependencies
-
-```bash
-glide update --strip-vendor
-```
+  Kedge vendors whole OpenShift and all its dependencies. (see comments in `glide.yaml` and `./scripts/vendor-openshift.yaml`)
+  It is possible that dependence you want to add is already in vendor as OpenShift dependency. If that is true, please make sure 
+  that you use the same version as OpenShift is using.
```
Comment (cdrage) [inline docs/development.md]: remove `as`

File: glide.lock [hunk 5]
```diff
@@ -1,231 +1,56 @@
+<<<<<<< HEAD
 hash: 6a587e17cdda6137bf312c6cf7f6981c0d42edd7ab71ea3eeebbd9d9a41fc86d
 updated: 2017-10-03T17:35:11.913154754+05:30
+=======
+hash: e96e18ffcf55c9537b8a7d121f8f35418aa73a0d0683132e5adc07253dddd23c
+updated: 2017-09-29T12:54:49.024337049+02:00
+>>>>>>> Add script to correctly vendor OpenShift
```
Comment (cdrage) [inline glide.lock]: uh oh

File: glide.lock [hunk 6]
```diff
@@ -1,231 +1,56 @@
...
 - name: github.com/spf13/viper
+<<<<<<< HEAD
   version: d9cca5ef33035202efb1586825bdbb15ff9ec3ba
 - name: github.com/ugorji/go
   version: ded73eae5db7e7a0ef6f55aace87a2873c5d2b74
   subpackages:
   - codec
+=======
+  version: 7fb2782df3d83e0036cc89f461ed0422628776f4
+>>>>>>> Add script to correctly vendor OpenShift
```
Comment (cdrage) [inline glide.lock]: more uh oh

File: glide.yaml [hunk 7]
```diff
@@ -1,33 +1,39 @@
...
-  - pkg/runtime
-  - pkg/util/intstr
+  version: 5215b55f46b2b919f50a1df0eaa5886afe4e3b3d
+
 - package: github.com/Sirupsen/logrus
   version: v0.7.3
+
+
+# Those are packages that OpenShift patches in its vendor directory.
+# Those patched libs are not published elswere.
```
Comment (cdrage) [inline glide.yaml]: elsewhere

File: glide.yaml [hunk 8]
```diff
@@ -1,33 +1,39 @@
...
+# See https://github.com/openshift/origin/pull/15797 for more info.
+# We have to ignore them here so they are not downloaded by glide.
+# We have to use modfied version of those packages from github.com/openshift/orogin/vendor,
```
Comment (cdrage) [inline glide.yaml]: use a modified

Comment (cdrage) [reply glide.yaml]: orogin -> origin

File: scripts/vendor-openshift.sh [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+    target_parent=`dirname $target`
+
+    rm -rf $target
+    mkdir -p $target_parent
+
+    echo "Moving $pkg from OpenShift vendor to this project vendor directory"
+    mv -f $TMP_OPENSHIFT/vendor/$pkg $target_parent    
+}
+
+# check if we cover everthing from openshift vendor
```
Comment (cdrage) [inline scripts/vendor-openshift.sh]: everything

File: scripts/vendor-openshift.sh [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+        echo "ERROR: structure for $domain is not defined"
+        exit 1
+    fi
+
+done
+
+
+# move packages from OpenShifts vendor dir to project vendor dir
+
+# for every package that is orgenized in two level directory strucure
```
Comment (cdrage) [inline scripts/vendor-openshift.sh]: organized

File: scripts/vendor-openshift.sh [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+for domain in $TWO_LEVEL; do
+    # on Linux you can use just find with  `-printf "$domain/%P\n"` instead of awk, but this is not available on MacOS
+    pkgs=`find -L "${TMP_OPENSHIFT}/vendor/$domain"  -maxdepth 1 -mindepth 1 -type d | awk -F/ '{ print($(NF-1)"/"$(NF)) }'`
+    for pkg in $pkgs; do
+        movePKG $pkg
+    done
+done
+
+# for every package that is orgenized in three level directory strucure
```
Comment (cdrage) [inline scripts/vendor-openshift.sh]: organized

File: Makefile [hunk 12]
```diff
@@ -109,3 +109,12 @@ test-unit-cover:
 	# go test doesn't support colleting coverage across multiple packages,
 	# generate go test commands using go list and run go test for every package separately
 	go list -f '"go test -race -cover -v -coverprofile={{.Dir}}/.coverprofile {{.ImportPath}}"' github.com/kedgeproject/kedge/...  | grep -v "vendor" | grep -v "e2e" | xargs -L 1 -P4 sh -c
+
+# Update vendoring
+# We have a bit messy vendoring now
```
Comment (cdrage) [inline Makefile]: `Vendoring is a bit messy right now`

File: glide.lock (hunk 6 above)
Comment (kadel) [reply glide.lock]: uh oh. ah, my bad. I had to do messy rebase. I will generate new glide.lock.
Thank you for noticing 👍

File: docs/development.md (hunk 1 above)
Comment (kadel) [reply docs/development.md]: fixed

File: docs/development.md (hunk 2 above)
Comment (kadel) [reply docs/development.md]: fixed

File: docs/development.md (hunk 3 above)
Comment (kadel) [reply docs/development.md]: fixed

File: docs/development.md (hunk 4 above)
Comment (kadel) [reply docs/development.md]: fixed

File: glide.yaml (hunk 7 above)
Comment (kadel) [reply glide.yaml]: fixed

File: glide.yaml (hunk 8 above)
Comment (kadel) [reply glide.yaml]: fixed

File: scripts/vendor-openshift.sh (hunk 9 above)
Comment (kadel) [reply scripts/vendor-openshift.sh]: fixed

File: scripts/vendor-openshift.sh (hunk 11 above)
Comment (kadel) [reply scripts/vendor-openshift.sh]: fixed

File: Makefile (hunk 12 above)
Comment (kadel) [reply Makefile]: fixed

Comment (kadel) [reply Makefile]: fixed

File: scripts/vendor-openshift.sh (hunk 11 above)
Comment (kadel) [reply scripts/vendor-openshift.sh]: fixed

File: glide.yaml (hunk 7 above)
Comment (kadel) [reply glide.yaml]: fixed

File: glide.yaml (hunk 8 above)
Comment (kadel) [reply glide.yaml]: fixed

File: scripts/vendor-openshift.sh (hunk 9 above)
Comment (kadel) [reply scripts/vendor-openshift.sh]: fixed

File: scripts/vendor-openshift.sh (hunk 10 above)
Comment (kadel) [reply scripts/vendor-openshift.sh]: fixed

**Comment by surajssd** [comment]
LGTM, merge whenever!

**Comment by containscafeine** [comment]
@cdrage merge away if you're good :)

**Comment by cdrage** [comment]
LGTM

**Pull request merged**